File: backend/src/main/java/cn/etarch/mao/harness/core/ContextManager.java
package cn.etarch.mao.harness.core;

import cn.etarch.mao.harness.llm.ChatRequest;
import cn.etarch.mao.harness.llm.LlmModelConfig;
import lombok.RequiredArgsConstructor;
import org.springframework.stereotype.Component;

import java.util.List;
import java.util.concurrent.atomic.AtomicBoolean;

/** 上下文窗口估算和会话交接压缩入口。 */
@Component
@RequiredArgsConstructor
public class ContextManager {

    private final TokenEstimator tokenEstimator;
    private final CompactionService compactionService;

    public int estimateTokens(List<ChatRequest.Message> messages) {
        return tokenEstimator.estimateMessages(messages);
    }

    public int estimateRequestTokens(ChatRequest request) {
        return tokenEstimator.estimateRequestTokens(request);
    }

    public CompactionService.SessionCompactionResult compactSession(
            Long sessionId, long expectedOldBoundary,
            List<PersistedChatMessage> messages, List<Long> snapshotMessageIds,
            ChatRequest normalRequest, LlmModelConfig modelConfig, CompactionConfig config,
            AgentEventListener listener, AtomicBoolean cancelFlag) {
        return compactionService.compactSession(sessionId, expectedOldBoundary, messages,
                snapshotMessageIds, normalRequest, modelConfig, config, listener, cancelFlag);
    }

    public List<ChatRequest.Message> prependSessionSummary(
            String summary, List<ChatRequest.Message> incrementalMessages) {
        return compactionService.prependSessionSummary(summary, incrementalMessages);
    }
}

File: backend-ts/src/harness/core/context-manager.ts
import type { ChatMessage, ChatRequest, LlmModelConfig } from '../llm/chat-request.js';
import type { CompactionConfig } from './compaction-config.js';
import type { CompactionService, SessionCompactionResult } from './compaction-service.js';
import type { TokenEstimator } from './token-estimator.js';
import type { AgentEventListener } from './agent-event-listener.js';
import type { PersistedChatMessage } from './persisted-chat-message.js';

export class ContextManager {
  constructor(
    private readonly tokenEstimator: TokenEstimator,
    private readonly compactionService: CompactionService,
  ) {}

  estimateTokens(messages: ChatMessage[]): number {
    return this.tokenEstimator.estimateMessages(messages);
  }

  estimateRequestTokens(request: ChatRequest): number {
    return this.tokenEstimator.estimateRequestTokens(request);
  }

  compactSession(
    sessionId: number | null,
    expectedOldBoundary: number,
    messages: PersistedChatMessage[],
    snapshotMessageIds: number[],
    normalRequest: ChatRequest,
    modelConfig: LlmModelConfig,
    config: CompactionConfig,
    listener: AgentEventListener | null,
    cancelFlag: { get(): boolean } | null,
    activeTokensHint?: number | null,
  ): Promise<SessionCompactionResult | null> {
    return this.compactionService.compactSession(
      sessionId, expectedOldBoundary, messages, snapshotMessageIds,
      normalRequest, modelConfig, config, listener, cancelFlag, activeTokensHint,
    );
  }

  prependSessionSummary(
    summary: string | null | undefined,
    incrementalMessages: ChatMessage[] | null,
    archiveHint?: string | null,
  ): ChatMessage[] {
    return this.compactionService.prependSessionSummary(summary, incrementalMessages, archiveHint);
  }
}
